mlir/lib/Dialect/Presburger/Set.cpp

```diff
@@ -72,6 +73,174 @@ void PresburgerSet::intersectSet(const PresburgerSet &set) {
+// subtraction and ~ denotes complement.
+// Let B be the basic set and S = (U_i S_i) be the set. We want B - (U_i S_i).
+//
+// Let S_i = /\_j S_ij. To compute B - S_i = B /\ ~S_i, we partition S_i based
+// on the first violated constraint:
+// ~S_i = (~S_i1) U (S_i1 /\ ~S_i2) U (S_i1 /\ S_i2 /\ ~S_i3) U ...
+// And the required result is (B /\ ~S_i1) U (B /\ S_i1 /\ ~S_i2) U ...
+// We recurse by subtracting U_{j > i} S_j from each of these parts and
+// returning the union of the results.
+//
+// TODO reimplement this heuristic:
+// As a heuristic, we try adding all the constraints and check if simplex
+// says that the intersection is empty. Also, in the process we find out that
+// some constraints are redundant, which we then ignore.
+void subtractRecursively(FlatAffineConstraints &B, Simplex &simplex,
+                         const PresburgerSet &S, unsigned i,
+                         PresburgerSet &result) {
+  if (i == S.getNumBasicSets()) {
+    FlatAffineConstraints BCopy = B;
+    // BCopy.simplify();
+    result.addFlatAffineConstraints(std::move(BCopy));
+    return;
+  }
+  const FlatAffineConstraints &S_i = S.getFlatAffineConstraints()[i];
+  auto initialSnap = simplex.getSnapshot();
+  // unsigned offset = simplex.numberConstraints();
+  // simplex.addBasicSetAsInequalities(set_i);
+  simplex.addFlatAffineConstraints(S_i);
+
+  if (simplex.isEmpty()) {
+    simplex.rollback(initialSnap);
+    subtractRecursively(B, simplex, S, i + 1, result);
+    return;
+  }
+
+  /*std::vector<bool> isMarkedRedundant;
+  for (size_t j = 0; j < 2 * S_i.getNumEqualities() + S_i.getNumInequalities();
+       j++)
+    isMarkedRedundant.push_back(simplex.isMarkedRedundant(offset + j));
+  */
+  simplex.rollback(initialSnap);
+  // TODO benchmark does it make a lot of difference if we always_inline this?
+  auto addInequalityFromEquality = [&](const ArrayRef<int64_t> &eq,
+                                       bool negated, bool strict) {
+    SmallVector<int64_t, 64> coeffs;
+    for (auto coeff : eq)
+      coeffs.emplace_back(negated ? -coeff : coeff);
+
+    // The constant is at the end
+    if (strict)
+      --coeffs[eq.size() - 1];
+
+    B.addInequality(coeffs);
+    simplex.addInequality(coeffs);
+  };
+  auto recurseWithInequalityFromEquality = [&, i](const ArrayRef<int64_t> &eq,
+                                                  bool negated, bool strict) {
+    size_t snap = simplex.getSnapshot();
+    addInequalityFromEquality(eq, negated, strict);
+
+    subtractRecursively(B, simplex, S, i + 1, result);
+
+    B.removeInequality(B.getNumInequalities() - 1);
+    simplex.rollback(snap);
+  };
+
+  for (size_t j = 0; j < S_i.getNumEqualities(); j++) {
+    // The first inequality is positive and the second is negative, of which
+    // we need the complements (strict negative and strict positive).
+    const auto &eq = S_i.getEquality(j);
+    /*if (isMarkedRedundant[2 * j]) {
```

Comment:
Do we need this for correctness or is this just an optimisation? In the later case I suggest that I first try to get a working end to end solution that enables testing.

Reply:
Unfortunately the libint code seems to be incorrect here on master. The not should be inverted here. In fact, I think this heuristic can be omitted until we show that it actually helps noticeably, as it seems to add a good amount of complexity to the code.

I suggest just treating equalities as a pair of inequalities for now, by replacing the body of the loop with:
```cpp
      const auto &eq = S_i.getEquality(j);
      recurseWithInequalityFromEquality(eq, true, true);
      addInequalityFromEquality(eq, false, false);
      recurseWithInequalityFromEquality(eq, false, true);
      addInequalityFromEquality(eq, true, false);
```

Reply:
Thanks for the feedback. I will adapt this and we can later on continue with this.